Issue/3836 big titles update

JorgeMucientes opened the pull request:
[**Issue/3836**](https://github.com/woocommerce/woocommerce-android/issues/3836)


### :goal_net: Goal
We use “large titles” for the four primary tab fragments, and we shrink the title as the user scrolls. The problem is that our first tab shows the name of the store, and the larger title means some store names are cut off.

This project aims to fix that by showing the name of the store in a smaller font under the “My Store” title, and the store name disappears as the user scrolls. Our designers have also decided that the tabs for the stats (“Today,” “Last Week,” etc.) shouldn’t appear until the user scrolls.

### :construction: How it's done?

Task is splitted into 2 main parts:
* **Add a subtitle** to "My Store" section that is only visible when the toolbar is expanded. On collapse we fade out the subtitle. he main changes this first part can be found in this [commit](https://github.com/woocommerce/woocommerce-android/pull/4196/commits/a3c6a1fcb97f38790fd17d9b2825f68a83ca4882). 
* **Date header** below the tabs will be remain sticky in portrait mode and scrollable in case of landscape. 

### 📱   Final demo

 
https://user-images.githubusercontent.com/2663464/122575684-498f7980-d051-11eb-921e-fb17bfccbc9b.mp4

(ignore leak canary notifications)

Commit:
Add subtitle for expanded collapsing toolbar state.

- Subtitle will fade on toolbar collapsed
- Added to MyStoreFragment.kt toolbar to display shop name

Commit:
Show tabs in MyStoreFragment.kt only when toolbar is collapsed

Commit:
Minor change so that subtitle is reproduces by android talk-back

Commit:
Revert accidental commited changes on Project.xml

Commit:
Ensure tabs are hidden if empty view is displayed

Commit:
Remove constant value to make the fade out effect on subtitle faster

Commit:
Remove unneeded AnimationExt.kt

- Expand and collapse animations already existed in the project

Commit:
Fix some issues with tab visibility in MyStoreFragment.kt

Commit:
Make MyStoreDateRangeView sticky in portrait and scrollable in landscape

- Extract MyStoreDateRangeView outside MyStoreStatsView to enable 2 different layouts of My Store fragment for portrait and landscape

Commit:
Remove unneeded findViewById()

Commit:
Set MyStoreDateRangeView.kt as MaterialCard

JorgeMucientes commented on .idea/codeStyles/Project.xml:
```diff
@@ -209,4 +209,4 @@
   </code_scheme>
-</component>
\ No newline at end of file
+</component>
```
?? unchanged, don't know why this is showing

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/BaseFragmentView.kt:
```diff
@@ -1,9 +1,6 @@
-     * Return the title that should appear in the action bar while this fragment is
-     * visible.
-     */
     fun getFragmentTitle(): String
```
Removing as method it's pretty self explanatory

Commit:
Remove duplicated code

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:
```diff
@@ -580,7 +605,7 @@ class MainActivity : AppUpgradeActivity(),
     private fun hasMagicLinkLoginIntent(): Boolean {
         val action = intent.action
         val uri = intent.data
-        val host = uri?.host?.let { it } ?: ""
```
unrelated to task. Removed `let` due to being redundant

nbradbury replied:
In general we like to avoid changes outside the current task, but at the same time we know it's hard not to quick-fix obvious warnings.

JorgeMucientes replied:
I understand your point, but how do you tackle this then? If you follow a strict policy of avoiding unrelated changes, when you see a safe quick-fix like this, what do you do? Do you once in a while create a task to do some "boy scouting" and cleaning small things? 

I don't have any problem reverting this if you consider it. Just curious about how you deal with this type of changes then.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreDateRangeView.kt:
```diff
@@ -14,8 +14,11 @@ import com.woocommerce.android.util.DateUtils
+    ctx: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : MaterialCardView(ctx, attrs, defStyleAttr) {
```
Ended up converting it into `MaterialCardView` just to get the proper background color for dark mode. Trickier than expected

Commit:
Remove unused colors

JorgeMucientes commented on WooCommerce/src/main/res/layout/fragment_my_store.xml:
```diff
@@ -101,6 +113,7 @@
                 android:id="@+id/stats_availability_image"
                 android:layout_width="wrap_content"
                 android:layout_height="wrap_content"
+                android:importantForAccessibility="no"
```
Empty case illustration should not require contentDescription so I'v set it as not important to remove unneeded warnings

JorgeMucientes commented on WooCommerce/src/main/res/layout/activity_main.xml:
```diff
@@ -29,6 +29,16 @@
+                    style="@style/Woo.TextView.Subtitle1.ToolbarSubtitle"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:importantForAccessibility="yes"
```
Tried to group content for talkbalk but it has been imposible to make talkback reproduce the subtitle unless I add this. In this case title and subtitle will be treated separately by talkback

nbradbury replied:
💯

JorgeMucientes commented on WooCommerce/src/main/res/layout-land/fragment_my_store.xml:
```diff
@@ -0,0 +1,135 @@
+                    android:visibility="gone"
+                    tools:visibility="visible" />
+
+                <!-- Date bar -->
+                <com.woocommerce.android.ui.mystore.MyStoreDateRangeView
+                    android:id="@+id/my_store_date_bar"
+                    style="@style/Woo.Card.WithoutPadding"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content" />
```
main change here, just setting the date range view as scrollable, the rest of the layout is unchanged

Commit:
Make tabs in My Store fragment sticky again

nbradbury commented:
@JorgeMucientes Just a heads-up that there are some merge conflicts here.

Commit:
Merge branch 'develop' into issue/3836-big-titles-update

# Conflicts:
#	.idea/codeStyles/Project.xml
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreDateRangeView.kt

Commit:
Move logic for getting subtitle value out of fragment to presenter

- Fix return count detekt issue in getSubtitle function

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStorePresenter.kt:
```diff
@@ -142,6 +142,15 @@ class MyStorePresenter @Inject constructor(
         }
     }
 
+    override fun getSelectedSiteName(): String? =
```
Good call on moving this logic to the presenter 👍

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:
```diff
@@ -463,6 +468,26 @@ class MainActivity : AppUpgradeActivity(),
+    }
+
+    private fun setFadingSubtitleOnCollapsingToolbar(subtitleView: TextView, subtitle: CharSequence) {
+        binding.collapsingToolbar.expandedTitleMarginBottom =
```
Does this margin need to be reset when there's no subtitle? If not, then does it need to be set every time here?

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:
```diff
@@ -463,6 +468,26 @@ class MainActivity : AppUpgradeActivity(),
+            resources.getDimensionPixelSize(dimen.expanded_toolbar_bottom_margin_with_subtitle)
+        subtitleView.text = subtitle
+        subtitleView.show()
+        binding.appBarLayout.addOnOffsetChangedListener(
```
Similar to the above comment, do we want to remove this listener when there's no subtitle? Because right now it still fires for the other tabs.

JorgeMucientes replied:
You're absolutely right! Good catch 🤦🏼 .  Refactored this in the last commit solving a couple issues

Not only the listener was not being unregistered but I was adding more than one listener for the same thing. Did the apropiate changes to only register a single listener when needed and properly unregister it when it corresponds. 

Did the same thing for the `expandedTitleMarginBottom`. Setting to 32dp when no subtitle present to replicate the exact behavior there was before adding the subtitle changes. Apparently the title in the collapsing toolbar has expandedTitleGravity="botton" by default. So if we change `expandedTitleMarginBottom` to 0dp the title in expanded mode was aligning to the bottom. To solve this I could either set expandedTitleGravity="center_vertical" always, which did almost did the trick, or set 32dp of `expandedTitleMarginBottom` when no subtitle. The second option was more accurate when the subtitle is not present. Check it out: 

(titles in the screenshots show the solution that was being applied)
![Screenshot 2021-06-19 at 23 20 36](https://user-images.githubusercontent.com/2663464/122656380-58ae1e80-d15a-11eb-94f3-f62fb945a335.png)

nbradbury commented:
@JorgeMucientes I completed my first round, and left only a few minor comments for you. If you make any changes over the weekend and I'm available, I'll follow up.

Commit:
Fix several issues with collapsing toolbar

- Properly adjust expandedTitleMarginBottom when no subtitle present
- Remove appBarOffsetListener when no subtitle and when activity is paused

JorgeMucientes commented:
BTW I'm getting a new error from detekt: 

`        LargeClass - 602/600 - [MainActivity] at /Users/jmucientes/Desktop/woocommerce-android/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:80:7`

I totally understand that is a code smell. MainActivty.kt seem to need a refactor as it has a decent size. But I feel that is out of scope for this task. On the other hand the code I've added to `MainActivity` it's basically setting values into binded views. It does not seem like code to fit in the presenter. I can extract my changes into a collaborator to reduce a few lines a turn off the detekt error but I think that is not the issue in MainActivity. Also I believe this rule is too strict I would maybe add this as a warning? 
WDYT?

nbradbury commented:
> BTW I'm getting a new error from detekt:

Ah, it's unfortunate that Detekt (which we very recently switched to) shows this error when it's not caused by your changes. Can you suppress it for that activity, with a `ToDo` note above it?

Commit:
Suppress LargeClass detekt error for MainActivity.kt

JorgeMucientes commented on config/detekt/baseline.xml:
```diff
@@ -76,6 +76,7 @@
     <ID>ForbiddenComment:ProductDetailRepository.kt$ProductDetailRepository$// TODO: add event to track sku availability success</ID>
     <ID>ForbiddenComment:RefundByItemsFragment.kt$RefundByItemsFragment$// TODO: Temporarily disabled, this will be used in a future release - do not remove</ID>
     <ID>ForbiddenComment:ShippingCarrierRatesViewModel.kt$ShippingCarrierRatesViewModel$// TODO: Once we start supporting countries other than the US, we'll need to verify what currency the shipping labels purchases use</ID>
+    <ID>ForbiddenComment:MainActivity.kt$MainActivity$// TODO: Extract logic out of MainActivity to reduce size and remove this @Suppress("LargeClass")</ID>
```
Kind of ironic that for suppressing "LargeClass" detekt error I had to add an additional "ignore rule" for the TODO comment as TODOs are also forbidden by detekt 😆

nbradbury replied:
![picard](https://user-images.githubusercontent.com/3903757/122754355-9053be00-d261-11eb-832a-26e753dc4001.gif)

nbradbury commented:
@JorgeMucientes Code changes look good, very nice work! I don't think anything has changed design-wise since you last posted a video, but just to be sure can you post a new video so @Garance91540 can approve the design?

JorgeMucientes commented:
Thanks for reviewing the code and your feedback 😃  

Sure! here is a video with the final result @Garance91540 

https://user-images.githubusercontent.com/2663464/122764913-40481c00-d2a0-11eb-98bf-6f37f0f8f470.mp4

Garance91540 commented:
Looking good! 👏👏

nbradbury reviewed (approved):
This all looks and works great! I'll ping our "Platform 9" team to get CI to run on this, and then I'm good to merge.

**Update:** @JorgeMucientes It turns out that in order to merge your changes, I have to pull them into my own branch and submit my own PR so CI can run. I've done this [here](https://github.com/woocommerce/woocommerce-android/pull/4241) and made sure you got the credit. Feel free to close this.
